File: ios/Features/Cookbook/BookmarkRibbon.swift
import SwiftUI

// MARK: - Bookmark Ribbon

/// Visual indicator for favorited recipes
/// Displays as a ribbon extending from the top-right corner of the page
struct BookmarkRibbon: View {
    @Environment(\.templateTokens) private var tokens

    // Configuration
    private let ribbonWidth: CGFloat = 32
    private let ribbonHeight: CGFloat = 56
    private let foldHeight: CGFloat = 12

    var body: some View {
        VStack(spacing: 0) {
            // Main ribbon body
            ribbonBody

            // Folded bottom triangle
            ribbonFold
        }
        .frame(width: ribbonWidth, height: ribbonHeight + foldHeight)
        .offset(y: -8) // Slightly overlap with page edge
    }

    // MARK: - Ribbon Body

    private var ribbonBody: some View {
        ZStack {
            // Ribbon shape
            Rectangle()
                .fill(tokens.palette.accent)
                .frame(width: ribbonWidth, height: ribbonHeight)

            // Heart icon
            Image(systemName: "heart.fill")
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(.white)
                .offset(y: 8)
        }
        .shadow(color: .black.opacity(0.2), radius: 2, x: -1, y: 2)
    }

    // MARK: - Ribbon Fold

    private var ribbonFold: some View {
        Path { path in
            path.move(to: CGPoint(x: 0, y: 0))
            path.addLine(to: CGPoint(x: ribbonWidth / 2, y: foldHeight))
            path.addLine(to: CGPoint(x: ribbonWidth, y: 0))
            path.closeSubpath()
        }
        .fill(tokens.palette.accent.opacity(0.7))
        .frame(width: ribbonWidth, height: foldHeight)
    }
}

// MARK: - Interactive Bookmark Ribbon

/// Bookmark ribbon with tap-to-toggle functionality
struct InteractiveBookmarkRibbon: View {
    @Binding var isFavorited: Bool
    @Environment(\.templateTokens) private var tokens

    // Animation state
    @State private var isAnimating = false

    private let ribbonWidth: CGFloat = 32
    private let ribbonHeight: CGFloat = 56
    private let foldHeight: CGFloat = 12

    var body: some View {
        Button(action: toggleFavorite) {
            VStack(spacing: 0) {
                // Main ribbon body
                ZStack {
                    Rectangle()
                        .fill(isFavorited ? tokens.palette.accent : tokens.palette.divider)
                        .frame(width: ribbonWidth, height: ribbonHeight)

                    Image(systemName: isFavorited ? "heart.fill" : "heart")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(isFavorited ? .white : tokens.palette.textSecondary)
                        .offset(y: 8)
                        .scaleEffect(isAnimating ? 1.3 : 1.0)
                }
                .shadow(color: .black.opacity(0.2), radius: 2, x: -1, y: 2)

                // Folded bottom
                Path { path in
                    path.move(to: CGPoint(x: 0, y: 0))
                    path.addLine(to: CGPoint(x: ribbonWidth / 2, y: foldHeight))
                    path.addLine(to: CGPoint(x: ribbonWidth, y: 0))
                    path.closeSubpath()
                }
                .fill((isFavorited ? tokens.palette.accent : tokens.palette.divider).opacity(0.7))
                .frame(width: ribbonWidth, height: foldHeight)
            }
            .frame(width: ribbonWidth, height: ribbonHeight + foldHeight)
            .offset(y: -8)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(isFavorited ? "Remove from favorites" : "Add to favorites")
    }

    private func toggleFavorite() {
        // Animate heart
        withAnimation(.spring(response: 0.3, dampingFraction: 0.5)) {
            isAnimating = true
        }

        // Toggle state
        withAnimation(.easeInOut(duration: 0.2)) {
            isFavorited.toggle()
        }

        // Reset animation
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
            withAnimation(.spring(response: 0.2, dampingFraction: 0.7)) {
                isAnimating = false
            }
        }
    }
}

// MARK: - Small Bookmark Badge

/// Smaller bookmark indicator for list views
struct BookmarkBadge: View {
    @Environment(\.templateTokens) private var tokens

    var body: some View {
        Image(systemName: "bookmark.fill")
            .font(.system(size: 12, weight: .semibold))
            .foregroundStyle(tokens.palette.accent)
            .padding(6)
            .background(
                Circle()
                    .fill(tokens.palette.accent.opacity(0.15))
            )
    }
}

// MARK: - Preview

#Preview("Bookmark Ribbon") {
    ZStack {
        Color(red: 0.95, green: 0.95, blue: 0.92)
            .ignoresSafeArea()

        RoundedRectangle(cornerRadius: 16)
            .fill(.white)
            .frame(width: 300, height: 400)
            .overlay(
                BookmarkRibbon()
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topTrailing)
                    .padding(.trailing, 24)
            )
            .shadow(radius: 8)
    }
}

#Preview("Interactive Ribbon") {
    struct PreviewWrapper: View {
        @State private var isFavorited = false

        var body: some View {
            ZStack {
                Color(red: 0.95, green: 0.95, blue: 0.92)
                    .ignoresSafeArea()

                RoundedRectangle(cornerRadius: 16)
                    .fill(.white)
                    .frame(width: 300, height: 400)
                    .overlay(
                        InteractiveBookmarkRibbon(isFavorited: $isFavorited)
                            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topTrailing)
                            .padding(.trailing, 24)
                    )
                    .shadow(radius: 8)

                Text("Tap ribbon to toggle")
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .offset(y: 220)
            }
        }
    }

    return PreviewWrapper()
}

#Preview("Bookmark Badge") {
    HStack(spacing: 20) {
        BookmarkBadge()
        Text("Grandma's Apple Pie")
    }
    .padding()
}

File: android/app/src/main/java/com/familyrecipe/features/cookbook/BookmarkRibbon.kt
package com.familyrecipe.features.cookbook

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bookmark
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.familyrecipe.designsystem.LocalTemplateTokens

// MARK: - Bookmark Ribbon

/**
 * Visual indicator for favorited recipes
 * Displays as a ribbon extending from the top-right corner of the page
 */
@Composable
fun BookmarkRibbon(
    modifier: Modifier = Modifier,
    ribbonWidth: Dp = 32.dp,
    ribbonHeight: Dp = 56.dp,
    foldHeight: Dp = 12.dp
) {
    val tokens = LocalTemplateTokens.current
    val accentColor = tokens.palette.accent
    val foldColor = tokens.palette.accent.copy(alpha = 0.7f)

    Box(
        modifier = modifier
            .size(width = ribbonWidth, height = ribbonHeight + foldHeight)
            .offset(y = (-8).dp) // Slightly overlap with page edge
            .drawBehind {
                // Main ribbon body
                drawRect(
                    color = accentColor,
                    topLeft = Offset.Zero,
                    size = Size(ribbonWidth.toPx(), ribbonHeight.toPx())
                )

                // Folded bottom triangle
                val foldPath = Path().apply {
                    moveTo(0f, ribbonHeight.toPx())
                    lineTo(ribbonWidth.toPx() / 2, ribbonHeight.toPx() + foldHeight.toPx())
                    lineTo(ribbonWidth.toPx(), ribbonHeight.toPx())
                    close()
                }
                drawPath(
                    path = foldPath,
                    color = foldColor
                )
            },
        contentAlignment = Alignment.TopCenter
    ) {
        // Heart icon
        Icon(
            imageVector = Icons.Filled.Favorite,
            contentDescription = "Favorited",
            modifier = Modifier
                .padding(top = 16.dp)
                .size(14.dp),
            tint = Color.White
        )
    }
}

// MARK: - Interactive Bookmark Ribbon

/**
 * Bookmark ribbon with tap-to-toggle functionality
 */
@Composable
fun InteractiveBookmarkRibbon(
    isFavorited: Boolean,
    onToggle: () -> Unit,
    modifier: Modifier = Modifier,
    ribbonWidth: Dp = 32.dp,
    ribbonHeight: Dp = 56.dp,
    foldHeight: Dp = 12.dp
) {
    val tokens = LocalTemplateTokens.current

    // Animation state
    var isAnimating by remember { mutableStateOf(false) }
    val scale by animateFloatAsState(
        targetValue = if (isAnimating) 1.3f else 1f,
        animationSpec = spring(
            dampingRatio = Spring.DampingRatioMediumBouncy,
            stiffness = Spring.StiffnessHigh
        ),
        finishedListener = { isAnimating = false },
        label = "heart_scale"
    )

    val ribbonColor = if (isFavorited) tokens.palette.accent else tokens.palette.divider
    val foldColor = ribbonColor.copy(alpha = 0.7f)
    val iconTint = if (isFavorited) Color.White else tokens.palette.textSecondary

    Box(
        modifier = modifier
            .size(width = ribbonWidth, height = ribbonHeight + foldHeight)
            .offset(y = (-8).dp)
            .clickable(
                interactionSource = remember { MutableInteractionSource() },
                indication = null
            ) {
                isAnimating = true
                onToggle()
            }
            .drawBehind {
                // Main ribbon body
                drawRect(
                    color = ribbonColor,
                    topLeft = Offset.Zero,
                    size = Size(ribbonWidth.toPx(), ribbonHeight.toPx())
                )

                // Folded bottom triangle
                val foldPath = Path().apply {
                    moveTo(0f, ribbonHeight.toPx())
                    lineTo(ribbonWidth.toPx() / 2, ribbonHeight.toPx() + foldHeight.toPx())
                    lineTo(ribbonWidth.toPx(), ribbonHeight.toPx())
                    close()
                }
                drawPath(
                    path = foldPath,
                    color = foldColor
                )
            },
        contentAlignment = Alignment.TopCenter
    ) {
        Icon(
            imageVector = if (isFavorited) Icons.Filled.Favorite else Icons.Outlined.FavoriteBorder,
            contentDescription = if (isFavorited) "Remove from favorites" else "Add to favorites",
            modifier = Modifier
                .padding(top = 16.dp)
                .size(14.dp)
                .scale(scale),
            tint = iconTint
        )
    }
}

// MARK: - Small Bookmark Badge

/**
 * Smaller bookmark indicator for list views
 */
@Composable
fun BookmarkBadge(
    modifier: Modifier = Modifier
) {
    val tokens = LocalTemplateTokens.current

    Surface(
        modifier = modifier,
        shape = CircleShape,
        color = tokens.palette.accent.copy(alpha = 0.15f)
    ) {
        Icon(
            imageVector = Icons.Filled.Bookmark,
            contentDescription = "Favorited",
            modifier = Modifier
                .padding(6.dp)
                .size(12.dp),
            tint = tokens.palette.accent
        )
    }
}
